File: xy4200/repo/xy4200/models/audit_log.py
from datetime import datetime
from app import db


class AuditLog(db.Model):
    __tablename__ = 'audit_log'

    id = db.Column(db.Integer, primary_key=True)
    
    action = db.Column(db.String(50), nullable=False, index=True)
    entity_type = db.Column(db.String(50), nullable=False, index=True)
    entity_id = db.Column(db.String(100), nullable=False, index=True)
    
    old_values = db.Column(db.Text)
    new_values = db.Column(db.Text)
    
    user = db.Column(db.String(100), index=True)
    ip_address = db.Column(db.String(50))
    user_agent = db.Column(db.String(500))
    
    timestamp = db.Column(db.DateTime, default=datetime.utcnow, index=True)
    
    notes = db.Column(db.Text)
    
    def to_dict(self):
        return {
            'id': self.id,
            'action': self.action,
            'entity_type': self.entity_type,
            'entity_id': self.entity_id,
            'old_values': self.old_values,
            'new_values': self.new_values,
            'user': self.user,
            'ip_address': self.ip_address,
            'user_agent': self.user_agent,
            'timestamp': self.timestamp.isoformat() if self.timestamp else None,
            'notes': self.notes
        }
    
    def __repr__(self):
        return f'<AuditLog {self.action}:{self.entity_type}:{self.entity_id}>'

File: xy11226/src/models/AuditLog.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const User = require('./User');
const Ticket = require('./Ticket');

const AuditLog = sequelize.define('AuditLog', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  action: {
    type: DataTypes.STRING(50),
    allowNull: false,
    comment: '操作类型：create, update, delete, classify, review, export, import'
  },
  module: {
    type: DataTypes.STRING(50),
    allowNull: false,
    comment: '模块：ticket, user, role, system'
  },
  recordId: {
    type: DataTypes.STRING(50),
    allowNull: true,
    comment: '记录ID'
  },
  operatorId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: User,
      key: 'id'
    },
    comment: '操作人ID'
  },
  oldValues: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: '旧值，JSON格式',
    get() {
      const rawValue = this.getDataValue('oldValues');
      return rawValue ? JSON.parse(rawValue) : null;
    },
    set(value) {
      this.setDataValue('oldValues', value ? JSON.stringify(value) : null);
    }
  },
  newValues: {
    type: DataTypes.TEXT,
    allowNull: true,
    comment: '新值，JSON格式',
    get() {
      const rawValue = this.getDataValue('newValues');
      return rawValue ? JSON.parse(rawValue) : null;
    },
    set(value) {
      this.setDataValue('newValues', value ? JSON.stringify(value) : null);
    }
  },
  ipAddress: {
    type: DataTypes.STRING(50),
    allowNull: true,
    comment: 'IP地址'
  },
  userAgent: {
    type: DataTypes.STRING(255),
    allowNull: true,
    comment: '用户代理'
  },
  remark: {
    type: DataTypes.STRING(500),
    allowNull: true,
    comment: '备注'
  }
}, {
  tableName: 'audit_logs',
  timestamps: true,
  updatedAt: false,
  indexes: [
    { fields: ['action'] },
    { fields: ['module'] },
    { fields: ['operatorId'] },
    { fields: ['createdAt'] }
  ]
});

AuditLog.belongsTo(User, { foreignKey: 'operatorId', as: 'operator' });
User.hasMany(AuditLog, { foreignKey: 'operatorId' });

module.exports = AuditLog;
